external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java

```diff
@@ -0,0 +1,292 @@
+    private final int maxUncommittedOffsets;
+    private final FirstPollOffsetStrategy firstPollOffsetStrategy;
+    private final KafkaSpoutStreams kafkaSpoutStreams;
+    private final KafkaSpoutTuplesBuilder<K, V> tuplesBuilder;
+    private final KafkaSpoutRetryService retryService;
+
+    private KafkaSpoutConfig(Builder<K,V> builder) {
+        this.kafkaProps = setDefaultsAndGetKafkaProps(builder.kafkaProps);
+        this.keyDeserializer = builder.keyDeserializer;
+        this.valueDeserializer = builder.valueDeserializer;
+        this.pollTimeoutMs = builder.pollTimeoutMs;
+        this.offsetCommitPeriodMs = builder.offsetCommitPeriodMs;
+        this.maxRetries = builder.maxRetries;
+        this.firstPollOffsetStrategy = builder.firstPollOffsetStrategy;
+        this.kafkaSpoutStreams = builder.kafkaSpoutStreams;
+        this.maxUncommittedOffsets = builder.maxUncommittedOffsets;
+        this.tuplesBuilder = builder.tuplesBuilder;
+        this.retryService = builder.retryService;
+    }
+
+    private Map<String, Object> setDefaultsAndGetKafkaProps(Map<String, Object> kafkaProps) {
+        // set defaults for properties not specified
+        if (!kafkaProps.containsKey(Consumer.ENABLE_AUTO_COMMIT)) {
+            kafkaProps.put(Consumer.ENABLE_AUTO_COMMIT, "false");
+        }
+        return kafkaProps;
+    }
+
+    public static class Builder<K,V> {
+        private final Map<String, Object> kafkaProps;
+        private Deserializer<K> keyDeserializer;
+        private Deserializer<V> valueDeserializer;
+        private long pollTimeoutMs = DEFAULT_POLL_TIMEOUT_MS;
+        private long offsetCommitPeriodMs = DEFAULT_OFFSET_COMMIT_PERIOD_MS;
+        private int maxRetries = DEFAULT_MAX_RETRIES;
+        private FirstPollOffsetStrategy firstPollOffsetStrategy = FirstPollOffsetStrategy.UNCOMMITTED_EARLIEST;
+        private final KafkaSpoutStreams kafkaSpoutStreams;
+        private int maxUncommittedOffsets = DEFAULT_MAX_UNCOMMITTED_OFFSETS;
+        private final KafkaSpoutTuplesBuilder<K, V> tuplesBuilder;
+        private final KafkaSpoutRetryService retryService;
+
+        /***
+         * KafkaSpoutConfig defines the required configuration to connect a consumer to a consumer group, as well as the subscribing topics
+         * The optional configuration can be specified using the set methods of this builder
+         * @param kafkaProps    properties defining consumer connection to Kafka broker as specified in @see <a href="http://kafka.apache.org/090/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html">KafkaConsumer</a>
+         * @param kafkaSpoutStreams    streams to where the tuples are emitted for each tuple. Multiple topics can emit in the same stream.
+         */
+        public Builder(Map<String, Object> kafkaProps, KafkaSpoutStreams kafkaSpoutStreams,
+                       KafkaSpoutTuplesBuilder<K,V> tuplesBuilder, KafkaSpoutRetryService retryService) {
```

Comment:
Would it be possible to have retryService be optional with a default instead of forceing everyone to pass in an argument.